google-cloud-core/src/main/java/com/google/cloud/ServiceOptions.java

```diff
@@ -476,6 +477,29 @@ static String getServiceAccountProjectId(String credentialsPath) {
     return project;
   }
 
+  /* 
+   * Returns true if the projectId is valid. This method checks whether the projectId
+   * contains only lowercase letters, digits and hyphens, starts with a lowercase letter
+   * and does not end with a hyphen, but does not check the length of projectId. This
+   * method is primarily used to protect against DNS hijacking.
+   */
+  static boolean isValidProjectId(String projectId) {   
+    for (char c : projectId.toCharArray()) {
+      if (!isLowerCase(c) && !isDigit(c) && c != '-') {
+        return false;
+      }
+    }
+    return projectId.length() > 0 && isLowerCase(projectId.charAt(0))
+      && !projectId.endsWith("-");
+  }
+
+  private static boolean isLowerCase(char c) {
```

Comment:
https://docs.oracle.com/javase/7/docs/api/java/lang/Character.html#isLowerCase(char)

Reply:
Please correct me if I am wrong. It seems Java's isLowerCase would return true for letters in languages other English, which is not allowed for project-id.

Reply:
Technically speaking, it's not about _English_, but non-ASCII characters--and furthermore, I don't see that in the references you mention, only "letters". What's actually allowed for a project ID?